cpp/pagerank_module/algorithm/pagerank.cpp

```diff
@@ -0,0 +1,140 @@
+#include <algorithm>
+#include <numeric>
+#include <queue>
+
+#include "pagerank.hpp"
+
+namespace pagerank_alg {
+
+PageRankGraph::PageRankGraph(const std::uint64_t number_of_nodes, const std::uint64_t number_of_edges,
+                             const std::vector<EdgePair> &edges)
+    : node_count_(number_of_nodes),
+      edge_count_(number_of_edges),
+      out_degree_(std::vector<std::uint64_t>(number_of_nodes)) {
+  pagerank_util::AdjacencyList<std::uint64_t> in_neighbours;
+  in_neighbours.Init(number_of_nodes);
+  for (const auto &[from, to] : edges) {
+    // Because PageRank needs a set of nodes that point to a given node.
+    in_neighbours.AddAdjacentPair(from, to);
+    out_degree_[from] += 1;
+  }
+  for (std::size_t node_id = 0; node_id < number_of_nodes; node_id++) {
+    const auto &adjacent_nodes = in_neighbours.GetAdjacentNodes(node_id);
+    for (const auto &adjacent_node : adjacent_nodes) {
+      ordered_edges_.emplace_back(adjacent_node, node_id);
+    }
+  }
+}
+
+std::uint64_t PageRankGraph::GetNodeCount() const { return node_count_; }
+
+std::uint64_t PageRankGraph::GetEdgeCount() const { return edge_count_; }
+
+const std::vector<EdgePair> &PageRankGraph::GetOrderedEdges() const { return ordered_edges_; }
+
+std::uint64_t PageRankGraph::GetOutDegree(const std::uint64_t node_id) const { return out_degree_[node_id]; }
+
+void CalculateOptimalBorders(const PageRankGraph &graph, const int number_of_threads,
```

Comment:
can number_of_threads be negative?

Reply:
`const std::uint64_t number_of_threads = std::thread::hardware_concurrency();`

You are right, above is the construction of the `number_of_threads` variable, which can't be negative. However, there should be a check whether is it zero(`0`) or not.  :+1: